### PR TITLE
Community categories reorder and persist open/close state

### LINKED ISSUE #12292: Add functionality for changing order of categories in community
# Feature Issue

leftover from https://github.com/status-im/status-react/issues/12066

### User Story

As a community admin, I want to change the order of categories

Designs
https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=16052%3A2354

### LINKED ISSUE #12293: Categories state (open/close) persistance
# Feature Issue

leftover from https://github.com/status-im/status-react/issues/12066

### User Story

As a community admin, I want to see categories in the same state they were left in the last session

### Implementation
changes in status-go required

### Designs
https://www.figma.com/file/aS1ct66VQ6V0cio7vSqS8UoG/Chat?node-id=16052%3A2354

### PR DESCRIPTION
**Pull request by Parveshdhull**
fixes #12292
fixes #12293

PR implements categories state persistence and functionality to reorder them
<div align="center">
Screenshot
</div>

### CODE REVIEW

File: src/status_im/communities/core.cljs [hunk 1]
```diff
@@ -586,3 +586,59 @@
...
+
+(defn category-hash [public-key community-id category-id]
+  (hash (str public-key community-id category-id)))
+
+(fx/defn store-category-state
+  {:events [::store-category-state]}
+  [{:keys [db]} community-id category-id state-open?]
+  (let [public-key (get-in db [:multiaccount :public-key])]
+    {::async-storage/set!
+     {(category-hash public-key community-id category-id) state-open?}}))
```
Comment (flexsurfer) [inline src/status_im/communities/core.cljs]: could we move the function to let ? it will be more readable then

Comment (flexsurfer) [reply src/status_im/communities/core.cljs]: and in that case we don't need category-hash function at all

File: src/status_im/communities/core.cljs [hunk 2]
```diff
@@ -586,3 +586,59 @@
...
+                     :on-error    #(log/error "failed to reorder community category" %)}]})
+
+(defn category-hash [public-key community-id category-id]
+  (hash (str public-key community-id category-id)))
+
+(fx/defn store-category-state
+  {:events [::store-category-state]}
+  [{:keys [db]} community-id category-id state-open?]
+  (let [public-key (get-in db [:multiaccount :public-key])]
+    {::async-storage/set!
```
Comment (flexsurfer) [inline src/status_im/communities/core.cljs]: state should be stored in status go, we store in async storage only setting for the entire app, this setting is per multiaccount

File: src/status_im/communities/core.cljs [hunk 3]
```diff
@@ -586,3 +586,59 @@
...
+  (let [public-key     (get-in db [:multiaccount :public-key])
+        categories-old (get-in db [:communities community-id :categories])
+        categories     (get
+                        (reduce (fn [acc category]
+                                  (let [category-id (get category 0)
+                                        hash        (get hashes category-id)
+                                        state       (get states hash)]
+                                    (assoc-in acc
+                                              [:categories category-id :state] state)))
+                                {:categories categories-old} categories-old)
```
Comment (flexsurfer) [inline src/status_im/communities/core.cljs]: can't it return the result? why do you need :categories key ?

**Review by flexsurfer — DISMISSED** [review]

File: src/status_im/communities/core.cljs (hunk 2 above)
Comment (flexsurfer) [reply src/status_im/communities/core.cljs]: discussed with Andrea, that's fine to use async storage, ignore my comment please

**Review by flexsurfer — APPROVED** [review]

**Comment by du82** [comment]
Can this also be done for channels on their own, not just categories?

**Comment by qoqobolo** [comment]
@Parveshdhull thanks for this PR!

### ISSUE 1: Slightly misleading message on the "Reorder categories" screen if there are no categories in the community
If no categories have been created in the community yet and a user opens the "Reorder categories" screen, the following message is shown:
`Your categories will appear here. To create a new category, click on the + button and select "Create category"`

However, there is no + button on the screen.
So it would be expected that the message was: 
`Your categories will appear here. To create a new category, go back to the community screen, click on the + button and select "Create category"`

Or the + button should be added to this screen as well.

https://user-images.githubusercontent.com/67952253/140049711-a78a2270-fec0-4131-a82b-130a9a6e9988.mp4

**Comment by Parveshdhull** [comment]
Hi @qoqobolo,

Yes, your version sounds much better.
Also, similar message is used for edit chats too. So, I will replace that one too 👍
Thanks

**Comment by qoqobolo** [comment]
@Parveshdhull thanks for the great work!
PR is tested and can be merged.

**Pull request merged**